processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -232,23 +248,39 @@ public CursorSequenceBuilder(
 
       final GenericColumn timestamps = index.getColumn(Column.TIME_COLUMN_NAME).getGenericColumn();
 
+      Iterable<Long> iterable = gran.iterable(interval.getStartMillis(), interval.getEndMillis());
+      if (descending) {
+        iterable = Lists.reverse(ImmutableList.copyOf(iterable));
+      }
+
       return Sequences.withBaggage(
           Sequences.map(
-              Sequences.simple(gran.iterable(interval.getStartMillis(), interval.getEndMillis())),
+              Sequences.simple(iterable),
               new Function<Long, Cursor>()
               {
                 @Override
                 public Cursor apply(final Long input)
                 {
                   final long timeStart = Math.max(interval.getStartMillis(), input);
-                  while (baseOffset.withinBounds()
-                         && timestamps.getLongSingleValueRow(baseOffset.getOffset()) < timeStart) {
-                    baseOffset.increment();
+                  final long timeEnd = Math.min(interval.getEndMillis(), gran.next(input));
+
+                  if (descending) {
+                    for (; baseOffset.withinBounds(); baseOffset.increment()) {
+                      if (timestamps.getLongSingleValueRow(baseOffset.getOffset()) < timeEnd) {
+                        break;
+                      }
+                    }
+                  } else {
+                    for (; baseOffset.withinBounds(); baseOffset.increment()) {
+                      if (timestamps.getLongSingleValueRow(baseOffset.getOffset()) >= timeStart) {
+                        break;
+                      }
+                    }
                   }
 
-                  long threshold = Math.min(interval.getEndMillis(), gran.next(input));
                   final Offset offset = new TimestampCheckingOffset(
-                      baseOffset, timestamps, threshold, maxDataTimestamp < threshold
+                      baseOffset, timestamps, timeStart, timeEnd,
+                      minDataTimestamp >= timeStart && maxDataTimestamp < timeEnd
```

Comment:
this seems like stronger condition than needs to be, will following suffice?
`
descending ? minDataTimestamp >= timeStart : maxDataTimestamp < timeEnd
`